### ISSUE TITLE
specify all dependencies in go.mod

### ISSUE THREAD

**Issue by ransomw1c** [oneconcern/datamon#312]
currently, some dependencies are specified in `.circleci/config.yaml`.

the task here is to keep track of all golang src dep info using the go module system.

----

`main go dependencies` are those in `go.mod`?

is it possible to change `go.mod` to install `gotestsum`, `moq`, and so on?

i'm viewing this as a desirable change based on the preference to maintain the linter rc instead of params passed to the linter in the circle-ci yaml.  the linter rc is conceptually equivalent to `go.mod` in this comparison wherein we're attempting to shrink `.circleci/config.yaml` in favor of project-global (i.e. not just ci) setup.

i can spin this comment off into a separate issue..  if the change is out of scope for this merge request, comment to that effect, and i'll do so.

_Originally posted by @ransomw1c in https://github.com/oneconcern/datamon/pull/305_

**Comment by fredbi** [comment]
@ransomw1c you have a point here.
Unfortunately the go module mechanism doesn't handle dev & build dependencies (like npm for instance). So basically, external tooling adds a lot of noise here.

I find the idea interesting, but having go.mod / go.sum essentially managed automatically by the builder makes retaining dependencies for such external tools not an easy thing.

The rule is that any added dep should successfully sustain a `go mod tidy`, which is a routinely used command.

**Comment by ransomw1c** [comment]
> like npm for instance

the set-it-and-forget-it nature of go modules does cause some conflicts with npm .. there are some comparisons, although i'm mostly viewing these in contrast for the moment -- i.e. i'm not thinking of go modules as being particularly like npm `dependencies`.  even in node, i'm usually not too particular about separating out `devDependencies`. 

imo, a comment in `go.mod` would suffice to separate the two (npm's `dependencies` and `devDependencies`) for now since

* (afaik) dependencies specified in `go.mod` actually get built into the binary unless the `imports` statements from `main.go` indicate that they're needed
* i'm pretty sure the tooling dependencies aren't a significant download cost, at least for the time being, compared to those required  to build the binary

**Comment by fredbi** [comment]
@ransomw1c sorry for the late on that topic.
The itching point is that go.mod is transparently maintained by all go build / go test commands.
Besides the `go mod tidy` command removes all unneeded dependencies.
So if, for instance, you don't have any import github.com/golangci/... anywhere, this dependency disappears
The download cost used to be significant in go1.12 (git clone) but is much better now since proxies are used by default and serve compressed archived of versioned repos.

My point wasn't really about download cost, but more about having deps which are not used by the repo, and not used by any importer of this repo.

Basically, those where the reasons supporting a separate `go get ...`  path in the CI jobs.

**Comment by fredbi** [comment]
I believe we are good now, with CI & Makefile ensuring build & test dependencies are resolved while keeping a minimal `go.mod`.

Unless some more specifically targeted issue pops up regarding dependencies management, we may close this one.

Notice that it is very important that test jobs update their dependencies from outside the freshly checked out repo, so as not to alter the `go.mod` while building (e.g. resulting in a dirty repo, which cannot be released).
